### ISSUE TITLE
Gradient accumulation

### ISSUE THREAD

**Issue by borgr** [davda54/sam#62]
How would gradient accumulation look like?
There are two calls to the loss currently, but I am not sure how it is done behind the scenes, and hence what is the right way to call it.
Normally accumulation would look like:

`loss =loss_function(output, model(input))
  accumulated_loss += loss
  self.accelerator.backward(loss)
  
  if (step_for_accumulation + 1) % gradient_accumulation_steps == 0 or step_for_accumulation == len(
          training) - 1:
      optimizer.step()
      lr_scheduler.step()
      optimizer.zero_grad()`
So there are several losses per one step, but with the first and second steps with the losses between them I am not sure what is the expected way to do that

**Comment by deanjgoldman** [comment]
Coincidentally, I had this same question. My best guess is just not zero-ing out the gradients after the first step. so if you were to follow one of the examples from the README, it would look like: `optimizer.first_step(zero_grad=False)`. @davda54 Is that about right? And thanks for making this repo!

**Comment by davda54** [comment]
Gradient accumulation is not possible with this implementation, as far as I know. The accumulated gradients are deleted between the two optimization steps. You would have to override the code to cache the gradients from the second step.
